src/Workspaces/Core/Desktop/Workspace/Host/TemporaryStorage/TemporaryStorageServiceFactory.ReferenceCountedDisposable.cs

```diff
@@ -0,0 +1,240 @@
+                        instanceToDispose = _instance;
+                    }
+
+                    // Ensure multiple calls to Dispose for this instance are a NOP.
+                    _instance = null;
+                }
+
+                instanceToDispose?.Dispose();
+            }
+
+            /// <summary>
+            /// Represents a weak reference to a <see cref="ReferenceCountedDisposable{T}"/> which is capable of
+            /// obtaining a new counted reference up until the point when the object is no longer accessible.
+            /// </summary>
+            public struct WeakReference
+            {
+                /// <summary>
+                /// DO NOT DISPOSE OF THE TARGET.
+                /// </summary>
+                private readonly WeakReference<T> _weakInstance;
+                private readonly StrongBox<int> _boxedReferenceCount;
+
+                public WeakReference(ReferenceCountedDisposable<T> reference)
+                    : this()
+                {
+                    if (reference == null)
+                    {
+                        throw new ArgumentNullException(nameof(reference));
+                    }
+
+                    var instance = reference._instance;
+                    var referenceCount = reference._boxedReferenceCount;
+                    if (instance == null)
+                    {
+                        // The specified reference is already not valid.
+                        return;
+                    }
+
+                    _weakInstance = new WeakReference<T>(instance);
+                    _boxedReferenceCount = referenceCount;
+                }
+
+                /// <summary>
+                /// Increments the reference count for the disposable object, and returns a new disposable reference to
+                /// it.
+                /// </summary>
+                /// <remarks>
+                /// <para>Unlike <see cref="ReferenceCountedDisposable{T}.TryAddReference"/>, this method is capable of
+                /// adding a reference to the underlying instance all the way up to the point where it is finally
+                /// disposed.</para>
+                ///
+                /// <para>The returned object is an independent reference to the same underlying object. Disposing of
+                /// the returned value multiple times will only cause the reference count to be decreased once.</para>
+                /// </remarks>
+                /// <returns>A new <see cref="ReferenceCountedDisposable{T}"/> pointing to the same underlying object,
+                /// if it has not yet been disposed; otherwise, <see langword="null"/> if the underlying object has
+                /// already been disposed.</returns>
+                public ReferenceCountedDisposable<T> TryAddReference()
+                {
+                    var weakInstance = _weakInstance;
+                    if (weakInstance == null || !_weakInstance.TryGetTarget(out var target))
```

Comment:
The !TryGetTarget is also implied by the actual implementation of TryAddReferenceImpl, not sure if that's an interesting fact or not.